design/one-pager-generic-usage-type.md

```diff
@@ -0,0 +1,397 @@
+
+- **Orphaned Managed Resources (MRs).** Consider a Helm Release MR that is
+deployed to an EKS Cluster MR. The Helm Release has a dependency on the EKS
+Cluster’s connection details - it uses them as Provider credentials. If the EKS
+Cluster is deleted before the Helm Release the Helm Release controller becomes
+unable to connect to its Provider in order to delete the Release. In practice
+the Release External Resource (ER) was implicitly deleted along with the GKE
+Cluster - there’s actually no work left for the Helm release controller to do -
+but it cannot know that. This results in the Helm Release MR being ‘orphaned’
+in a pending deletion that can never be satisfied.
+
+- **Orphaned ‘Side-Effect’ Resources.** Consider the same example as above, and
+assume the Helm Release includes a Kubernetes Service resource of type:
+LoadBalancer. A side-effect of the creation of this Service will be to create an
+ELB in the EKS cluster’s VPC. This ELB is not managed by or even known to
+Crossplane, and is not deleted along with the EKS cluster. This means that the
+Helm Release (and therefore its Service) must be deleted before the EKS cluster
+in order to ensure the EKS cluster’s controllers trigger deletion of the ELB.
+If it is not, the ELB will be orphaned, costing money and blocking the deletion
+of the VPC it was created in.
+
+These problems are exacerbated by the fact that Kubernetes uses "background
+cascading deletion" by default. Put otherwise, the successful deletion of a
+parent resource is not dependent on the successful deletion of its children.
+Instead, the parent resource will be deleted successfully and the children
+(i.e. resources that declare the parent as an owner reference) will later be
+deleted by the Kubernetes garbage collector. This makes it difficult to diagnose
+or even notice resources that become orphaned when their parent XR is deleted.
+
+## Goals
+
+- Support "cross-provider" dependencies, for example an MR from provider-helm
+depending on a ProviderConfig from provider-aws.
+- Don’t require Crossplane users to manage fine-grained RBAC access.
+- Don’t grant Crossplane providers broad RBAC access (e.g. to all other providers).
+
+## Proposal
+
+We propose introducing a new `Usage` type that can be used to declare usage
+relationships between Crossplane resources. This type will be defined by the
+Core Crossplane repository, and be available for use as part of a composition.
+
+The relations defined by this type will be enforced by an admission webhook that
+will be running as part of the Crossplane core. Similar to the upstreams [liens
+proposal], the webhook will reject deletions of resources that are in use by
+other resources with a ["409 Conflict" error code].
+
+### API
+
+We will introduce a new `Usage` type as follows:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Usage
+metadata:
+  name: release-uses-cluster
+spec:
+  of:
+  - apiVersion: eks.upbound.io/v1beta1
```

Comment:
Clarify that such a reference applies to every group version of that kind. E.g. a usage of `eks.upbound.io/v1beta1` clusters implies a usage of `eks.upbound.io/v1` clusters.

Reply:
Another details, not sure I saw that specified anywhere: what do dangling references mean where the GVK is unknown to discovery (rest mapper)?

Reply:
> Clarify that such a reference applies to every group version of that kind.

FWIW this is pretty much a given for all of our references. We can be explicit here, but it's implicit in 99% of our other designs and APIs. 🙂 I believe API conventions now recommend a versionless pattern for such references, but most of our APIs use this older convention.

Reply:
One thing that I don't like with the [current convention](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#multiple-resource-reference) is, users would need to provide group & resource which is less intuitive than apiVersion & kind. The latter could easily be copied/pasted from the object to be referenced but when we use group & resource, they would need to make a conversion, e.g. `Object` -> `objects` and `kubernetes.crossplane.io/v1alpha1` -> `kubernetes.crossplane.io`

Reply:
Can we do the conversion internally - remove version from group and convert kind to resource?